Interfaces Enhanced

XedinUnknown opened the pull request:
- Interfaces enhanced to be more granular and semantic.
- Moved exceptions to own namespace.
- Interop increased.
- Tests fixed and improved.
- Docs improved too!

XedinUnknown committed:
Placed exceptions into a sub-namespace

XedinUnknown committed:
Tests up to standard

XedinUnknown committed:
Code standards fixed

XedinUnknown committed:
Improved throws description

XedinUnknown committed:
Test brought up to standrd

XedinUnknown committed:
Shifted the func tests to be unit tests

XedinUnknown committed:
Added `FactoryAwareInterface`

XedinUnknown committed:
Now also depending on the exception interface package

And Composer will prefer stable packages.

XedinUnknown committed:
Exceptions now implement `ThrowableInterface

XedinUnknown committed:
Added dev dependency on stringable interface

XedinUnknown committed:
Removed redundant mocked method

XedinUnknown committed:
Fixed missing import

XedinUnknown committed:
Factory now allowed to create anything, not only objects

This fixes #3.

XedinUnknown committed:
`FactoryAwareInterface#getFactory()` now allowed to return `null`

XedinUnknown committed:
Removed redundant implementations and tests for `ThrowableInterface`

XedinUnknown committed:
Exceptions are now factory-aware

XedinUnknown committed:
`make()` no longer receives `$id` parameter

XedinUnknown committed:
`make()` now also accepts a container as configuration

XedinUnknown committed:
Now using `precise` distro to build

mecha reviewed (changes requested):
Love it. I only have 1 question:

Does `CouldNotMakeExceptionInterface` need `getSubjectId()` now that the factory no longer creates using an ID? I'm thinking it doesn't. I can see how some form of "ID" can be useful when catching exceptions but `getSubjectConfig()` is available, so any form of identification can come from there.

What do you think?

XedinUnknown commented:
Good catch! Absolutely correct!

XedinUnknown committed:
`getSubjectId()` is no longer required due to absence of subject ID

XedinUnknown committed:
Fixed return type of `getSubejectConfig()`

Since the config can be an array or a container, this must match.

This concludes and fixes #2.

mecha reviewed (approved):
LGTM
